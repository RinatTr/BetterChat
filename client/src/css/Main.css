div.main-container {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
}

div.messages-container {
  margin-left: 15px;
}
div.form-container {
  width: 100%;
  position: fixed;
  bottom: 0px;
  border: none;
  border-top: 1.3px solid #55D3B4;
  padding-top: 2px;
  background: white;
}

ul {
  list-style-type: none;
  margin-block-start: 0em;
  margin-block-end: 5em;
  padding-inline-start: 0px;
  display: flex;
  flex-direction: column-reverse;
  /* overflow: scroll; */

}

form {
  display: grid;
  grid-template-columns: 15fr 1fr;
}

form input {
  border: none;
  padding: 15px;
  font-family: 'Roboto', sans-serif;
  font-weight: 700;
  font-size: calc(6px + 1vmin);
}

form input::placeholder {
  color: #C2CCCC;
}

button {
  border: none;
  background: none;
}

button i.material-icons {
  font-size: calc(20px + 2vmin);
  color: lightgrey;
}

