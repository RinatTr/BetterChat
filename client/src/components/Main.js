import React, { Component } from 'react';
import Message from './Message.js';
import * as Util from '../util/util.js';
import '../css/Main.css';
import { w3cwebsocket as W3CWebSocket } from "websocket";
const client = new W3CWebSocket('ws://localhost:8000');

class Main extends Component {
    constructor(props) {
        super(props)
        this.state = {
            msgs: [],
            username: "",
            user_id: null,
            prompt: "",
            invalid_user: false,
            non_unique: false
        }
        this.handleSubmit = this.handleSubmit.bind(this);
    }
    async componentDidMount() {
        client.onopen = () => {
            console.log('WebSocket Client Connected');
        };
        client.onmessage = (message) => {
            setTimeout(function() { Util.updateMessages(this) }.bind(this), 1000)
            if (typeof message.data === 'string') {
                console.log("Received: '" + message.data + "'");
            }
        };
        Util.updateMessages(this);  
    }

    handleChange = (e) => {
        this.setState({prompt: e.target.value})
    }
    async handleSubmit(e) {
        e.preventDefault();
        let { prompt } = this.state;
        //if no username, set username. else, post message
        client.send(prompt)
        if (!this.state.username) {
            let username = prompt;
            if (Util.isNotLongUsername(username)) {
                Util.updateUser(this, username);
            } else {
                this.setState({invalid_user: true})
            }            
        } else {
            let msg = { body: prompt,
                        user_id: this.state.user_id };
            Util.updateMessages(this, msg); 
        }
    }

    render() {
        let { msgs, prompt, username, invalid_user, non_unique } = this.state; 
        return (
            <div className="main-container">
                <div className="messages-container">
                    <ul>
                        {msgs.map((msg, i) => {
                            return <Message 
                                key={i}
                                body={msg.body} 
                                username={msg.username}
                                msg_timestamp={msg.created_at}
                                user_timestamp={msg.user_created_at}
                                />
                            })}
                    </ul>
                </div>
                <div className="form-container">
                    {invalid_user || non_unique ? <span className="invalid">{Util.invalidCaseStr(invalid_user, non_unique)}</span> : ""}
                    <form onSubmit={this.handleSubmit}>
                        <input 
                            name="prompt" 
                            onChange={this.handleChange} 
                            value={prompt} 
                            placeholder={username ? "what's happening?" : "Enter a username"}
                        />
                        <button type="submit"><i className="material-icons">{username ? "send" : "account_circle"}</i></button>
                    </form>
                </div>
            </div>
        )
    }
}

export default Main;