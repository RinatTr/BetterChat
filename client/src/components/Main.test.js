//When comp rerenders, need to be booleans. 
import React from 'react';
import renderer from 'react-test-renderer';
import Main from './Main';
import axios from 'axios';

// Jest does not use proxy for API calls
const PROXY = "http://localhost:3100";

test('Main snapshot test', () => {
  const component = renderer.create(<Main />);
  const tree = component.toJSON();
  console.log(tree.props)
  expect(tree).toMatchSnapshot();
});

test('non-unique username request fails with expected error code', () => {
    const EXISTING_USERNAME = "wolverine";
    return axios.post(`${PROXY}/api/users/${EXISTING_USERNAME}`).catch(e => {
        let err_code = e.response.data.error.code; 
        expect(err_code).toBe('23505');
    });
  });