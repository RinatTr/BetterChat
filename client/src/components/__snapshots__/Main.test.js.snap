// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main snapshot test 1`] = `
<div
  className="main-container"
>
  <div
    className="messages-container"
  >
    <ul />
  </div>
  <div
    className="form-container"
  >
    
    <form
      onSubmit={[Function]}
    >
      <input
        name="prompt"
        onChange={[Function]}
        placeholder="Enter a username"
        value=""
      />
      <button
        type="submit"
      >
        <i
          className="material-icons"
        >
          account_circle
        </i>
      </button>
    </form>
  </div>
</div>
`;
